### PR TITLE
[Feature] Same version computed on different branches

### LINKED ISSUE #3453: [Feature] Same version computed on different branches
**Describe the bug**
Hello, to be honest i'm not sure if it is a gitversion issue or i'm not using gitflow right. But i hope you could help me.

I created this repository https://github.com/xpicio/compute-version to test incremental version using conventional commits, the repo is based on gitflow model and i'm using git tower with the gitflow plugin.

In the actual state (after i closed an hotfix, pushed both branches develop and main, and finally added a tag) when i compute the version on develop or main branch, the version is the same: 0.2.4. This behaviour is causing an issue in CI/CD pipelines, when i need to restart a pipeline (should happen to fix a bug on the pipeline) the artifacts tagged from the develop branch will assume the same version of the artifacts from the main branch.

## Expected Behavior
| action         | branch  | commit message | sha                                      | version       |
| -------------- | ------- | -------------- | ---------------------------------------- | ------------- |
| close hotfix (random) |               |                                   |                                          |                |
|                | develop |                | 53720f5d878cbc9eab27fbb127615c3e85971409 | 0.3.0-alpha.9 |
|                | main    |                | 53720f5d878cbc9eab27fbb127615c3e85971409 | 0.2.4         |
| new tag v0.2.4 |         |                | 53720f5d878cbc9eab27fbb127615c3e85971409 |               |
|                | develop |                | 53720f5d878cbc9eab27fbb127615c3e85971409 | 0.x.x-alpha.x         |
|                | main    |                | 53720f5d878cbc9eab27fbb127615c3e85971409 | 0.2.4         |

The version on develop branch is different to main

## Actual Behavior
| action         | branch  | commit message | sha                                      | version       |
| -------------- | ------- | -------------- | ---------------------------------------- | ------------- |
| close hotfix (random) |               |                                   |                                          |                |
|                | develop |                | 53720f5d878cbc9eab27fbb127615c3e85971409 | 0.3.0-alpha.9 |
|                | main    |                | 53720f5d878cbc9eab27fbb127615c3e85971409 | 0.2.4         |
| new tag v0.2.4 |         |                | 53720f5d878cbc9eab27fbb127615c3e85971409 |               |
|                | develop |                | 53720f5d878cbc9eab27fbb127615c3e85971409 | 0.2.4         |
|                | main    |                | 53720f5d878cbc9eab27fbb127615c3e85971409 | 0.2.4         |

The version on develop branch is equal to main

## Steps to Reproduce
```
docker run --rm gittools/gitversion:6.0.0-beta.1 \
    /url https://github.com/xpicio/compute-version.git \
    /b develop
```

```
docker run --rm gittools/gitversion:6.0.0-beta.1 \
    /url https://github.com/xpicio/compute-version.git \
    /b main
```

## Your Environment
- Version Used: 5.12.0, 6.0.0-beta.1
- Operating System and version: macOs Ventura 13.2.1
- Link to your project: https://github.com/xpicio/compute-version

### PR DESCRIPTION
**Pull request by HHobeck**
[Feature] Same version computed on different branches:
- #3453

Close #3453

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by HHobeck** [comment]
Hi @asbjornu. I'm not sure if you are agree with the new branch related property with name `take-incremented-version`. Please take a look and give me feedback.

Thank's

File: docs/input/docs/reference/configuration.md [hunk 1]
```diff
@@ -52,16 +53,22 @@ commit-date-format: yyyy-MM-dd
...
+strategies:
+- ConfiguredNextVersion
+- MergeMessage
+- TaggedCommit
+- TrackReleaseBranches
+- VersionInBranchName
 branches:
   develop:
-    mode: ContinuousDeployment
+    take-incremented-version: TakeAlwaysIncrementedVersion
```
Comment (asbjornu) [inline docs/input/docs/reference/configuration.md:64]: Hm, I think `take-incremented-version: TakeAlwaysIncrementedVersion` have room for improvements. Give me some time to think through the use case and see if I can come up with better names.

**Review by asbjornu — CHANGES_REQUESTED** [review]

**Comment by asbjornu** [comment]
Do I understand this issue correctly when it's about whether we want to consider the `git tag` of the current commit or not? If so, perhaps `ignore-tag: False` would be more intuitive?

**Comment by HHobeck** [comment]
> Do I understand this issue correctly when it's about whether we want to consider the `git tag` of the current commit or not? If so, perhaps `ignore-tag: False` would be more intuitive?

At the end the criteria you are mentioned is correct. But you need to think about the semantic meaning. We have one or more strategies who are determining the existing base versions with a indication should the base increment yes or no. We are calculating based on this the next version and take the one which is highest. The result is we have exactly one base version and one incremented version (which might be equal). Now this feature comes in place where we want for some branches not to use the incremented version but the base version. 

I can think of the following use cases:
- Take always the base version and not the incremented version,
- Take the base version when the current commit was released otherwise the incremented version,
- Take always the incremented version and not the base version

Your proposed property with name `ignore-tag` does not reflect this semantic meaning if you ask me. It's also not about ignoring tags because the current tag will be considered and used to calculate the incremented version. If someone uses the `ignore-tag` property he could expect that all tags are ignored on the branch and not considered for the version calculation (like you would specify to ignore commits).

Does it make sense?

**Comment by HHobeck** [comment]
> I can think of the following use cases:
> 
> - Take always the base version and not the incremented version,
> - Take the base version when the current commit was released otherwise the incremented version,
> - Take always the incremented version and not the base version

While writing this I came up with the problem that we needs to extend the criteria e.g. for trunk based workflows. Because on trunk based workflows the released version on main will be not always marked with a tag (it's optional). So the criteria might:

- The current commit was tagged or the current branch IsMain (at least when trunk-based stratregy was used)

**Comment by asbjornu** [comment]
Right. But wouldn't then `increment: None` work, since for the specific branch we don't want to increment either `Major`, `Minor` or `Patch`?

**Comment by HHobeck** [comment]
> Right. But wouldn't then `increment: None` work, since for the specific branch we don't want to increment either `Major`, `Minor` or `Patch`?

But we want to increment the version on the branch but only if the commit was not released (marked with a tag).

**Comment by HHobeck** [comment]
Maybe the following enum values are better:

```csharp
internal enum TakeIncrementedVersion
{
    Never,
    IfNotReleased/Tagged,
    Always
}
```

File: docs/input/docs/reference/configuration.md (hunk 1 above)
Comment (asbjornu) [reply docs/input/docs/reference/configuration.md:64]: Let's do as discussed on Slack and group this and `prevent-increment-of-merged-branch-version` into a new top-level property as such:

```yaml
prevent-increment:
    when-versioned-branch-merged: true
    when-tagged: true
```

**Comment by HHobeck** [comment]
> ```yaml
> prevent-increment:
>     when-versioned-branch-merged: true
>     when-tagged: true
> ```

Actually I would like to do this in a separate PR if you not mind. Please review the changes and merge it to main. Thank you.

**Comment by HHobeck** [comment]
I think this PR can be merged right!?

File: src/GitVersion.Core.Tests/IntegrationTests/DevelopScenarios.cs [hunk 2]
```diff
@@ -46,7 +46,7 @@ public void WhenDevelopBranchedFromTaggedCommitOnMainVersionDoesNotChange()
         using var fixture = new EmptyRepositoryFixture();
         fixture.MakeATaggedCommit("1.0.0");
         fixture.BranchTo("develop");
-        fixture.AssertFullSemver("1.0.0");
+        fixture.AssertFullSemver("1.1.0-alpha.0");
```
[the remark this answers is not in the document]
Comment (asbjornu) [reply src/GitVersion.Core.Tests/IntegrationTests/DevelopScenarios.cs:49]: Indeed! :)

**Review by asbjornu — APPROVED** [review]

**Pull request merged**